hal/src/peripherals/adc/config.rs

```diff
@@ -0,0 +1,171 @@
+///
+/// Therefore, the overall formula for calculating Sample rate (SPS) can be
+/// calculated like so:
+///
+/// ## For single sample
+/// ```
+/// SPS = (GCLK_ADC / clk_divider) / (sample_clock_cycles + bit_width)
+/// ```
+/// ## For multiple samples (Averaging or Summed)
+/// ```
+/// SPS = (GCLK_ADC / clk_divider) / (n * (sample_clock_cycles + bit_width))
+/// ```
+#[derive(Copy, Clone)]
+pub struct Config {
+    pub clk_divider: Prescaler,
+    pub sample_clock_cycles: u8,
+    pub bit_width: Resolution,
+    pub accumulation: Accumulation,
+    pub vref: Reference,
+}
+
+impl Config {
+    ///
+    /// Configure the ADC to sample at 250_000 SPS (Assuming the clock source is
+    /// 48MHz) using the following settings:
+    /// * clock divider factor of 32
+    /// * 6 clock cycles per sample
+    /// * 12bit sampling
+    /// * Single accumulation (No averaging or summing)
+    ///
+    /// ## Additional reading settings by default
+    /// * Use VDDANA as reference voltage for a full 0.0-3.3V reading
+    pub fn new() -> Self {
+        Self {
+            clk_divider: Prescaler::Div32,
+            sample_clock_cycles: 6,
+            bit_width: Resolution::_12bit,
+            accumulation: Accumulation::Single,
+            vref: Reference::Intvcc1,
+        }
+    }
+
+    ///
+    /// This setting adjusts the ADC clock frequency by dividing the input clock
+    /// for the ADC.
+    ///
+    /// ## Example:
+    /// * Input clock 48MHz, div 32 => ADC Clock is 1.5MHz
+    pub fn clock_divider(mut self, div: Prescaler) -> Self {
+        self.clk_divider = div;
+        self
+    }
+
+    /// This setting adjusts the bit width of each ADC sample
+    pub fn sample_resolution(mut self, bit_width: Resolution) -> Self {
+        self.bit_width = bit_width;
+        self
+    }
+
+    pub fn with_vref(mut self, reference: Reference) -> Self {
+        self.vref = reference;
+        self
+    }
+
+    /// Sets how the ADC will accumulate values before actually returning a
+    /// value.
+    ///
+    /// The default is single (ADC will return a sample as soon as it is
+    /// measured)
+    ///
+    /// Setting [AdcAccumulation::Summed] will make the ADC take 'n' samples,
+    /// and sum the total before returning it
+    ///
+    /// Setting [AdcAccumulation::Average] will make the ADC take 'n' samples,
+    /// and average the total before returning it
+    ///
+    /// NOTE: Selecting [AdcAccumulation::Summed] or [AdcAccumulation::Average]
+    /// will reduce the overall ADC sample rate by a factor of 1/n, and the
+    /// returned value will be 16bits long no matter what the sample Bit
+    /// width was selected as
+    pub fn accumulation_method(mut self, method: Accumulation) -> Self {
+        self.accumulation = method;
+        self
+    }
+
+    /// This adjusts the number of ADC clock cycles taken to sample a single
+    /// sample. The higher this number, the longer it will take the ADC to
+    /// sample each sample.
```

Comment:
What is the maximum value, and assuming it's not 8b what happens if a config is specified that's over it?